### PR TITLE
DietPi-Software | Moonlight: Game streaming client for Sunshine and NVIDIA GameStream

### PR DESCRIPTION
**Pull request by mtekman**
Add [Moonlight](https://github.com/moonlight-stream/moonlight-embedded/) game streaming client

This is the embedded version suitable for RPI, based on their debian bullseye build.

The installation and deinstallation script is based on the [setup script](https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/setup.deb.sh) shown on their [packages page](https://github.com/moonlight-stream/moonlight-embedded/wiki/Packages).

In terms of documentation, I'm not sure what's really needed. People can just run `moonlight list` and it will automatically look for any broadcasting services and try to match and pair, as given in their [Usage section](https://github.com/moonlight-stream/moonlight-embedded/wiki/Usage) on their wiki.

### CODE REVIEW

**Comment by MichaIng** [comment]
Many thanks for your contribution. I'll go though it in detail in a few hours. Does it internally use `pip` for something or why do you install it explicitly? The official installer script does not do that.

**Comment by mtekman** [comment]
Hey, good question. From the `detect_os_modern_python` function in their script it checks for pip, but I believe this is purely to detect the OS, so it's likely not necessary. I will drop it.

**Comment by mtekman** [comment]
One thing I'm not sure about is that moonlight only works if `dtoverlay=vc4-fkms-v3d` is set in config.txt (and then after a reboot).

I've seen several mentions to modifying this via `G_CONFIG_INJECT` during installation, but I'm not sure how to "unset" it back to what it was during uninstallation.

File: dietpi/dietpi-software [hunk 1]
```diff
@@ -966,6 +966,12 @@ Available commands:
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,1]=0
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,2]=0
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,10]=0
+		#------------------
+		software_id=206
+		aSOFTWARE_NAME[$software_id]='Moonlight'
+		aSOFTWARE_DESC[$software_id]='Game streaming client for Sunshine and NVIDIA GameStream'
+		aSOFTWARE_CATX[$software_id]=5
+		aSOFTWARE_DOCS[$software_id]='https://github.com/moonlight-stream/moonlight-embedded/wiki'
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Let's use the URL to our docs even that the section needs to be created first:
```suggestion
		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/gaming/#moonlight'
```

File: dietpi/dietpi-software [hunk 2]
```diff
@@ -11103,6 +11109,34 @@ _EOF_
...
 		fi
 
+		if To_Install 206 # Moonlight
+		then
+			## Install script adapted from
+			## - https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/setup.deb.sh
+
+			# APT deps
+			aDEPS=('apt-transport-https' 'ca-certificates' 'debian-keyring' 'debian-archive-keyring' 'gnupg')
+
```
Comment (MichaIng) [inline dietpi/dietpi-software]: None of these is neeeded:
```suggestion
```
- APT has native HTTPS support, so `apt-transport-https` is just a transitional dummy package since Buster: https://packages.debian.org/apt-transport-https
- `ca-certificates` is necessarily pre-installed, otherwise the script would not even run until here 😉.
- `debian-keyring` is for Debian package maintainers to upload packages, never for users/admins to install packages: https://packages.debian.org/bookworm/debian-keyring
- `debian-archive-keyring` is a hard dependency of APT on Debian, while on Raspbian it is `raspbian-archive-keyring`. So the correct one is always pre-installed: https://packages.debian.org/bullseye/apt
- `gnupg` is as well pre-installed on DietPi, otherwise a bunch of other tasks and installs wouldn't work either.

File: dietpi/dietpi-software [hunk 3]
```diff
@@ -11103,6 +11109,34 @@ _EOF_
...
+			# APT key
+			local base_url="https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded"
+			local gpg_url="$base_url/gpg.5AEE46706CF0453E.key"
+			G_CHECK_URL "$gpg_url"
+			G_EXEC eval "curl -1sLf '$gpg_url' | gpg --dearmor > '$gpg_keyring_path'"
+
+			# APT list
+			local repo_url="$base_url/deb/raspbian"
+			G_EXEC eval "echo 'deb [signed-by=$gpg_keyring_path] $repo_url $G_DISTRO_NAME main' > $repo_path"
+			G_EXEC eval "echo 'deb-src [signed-by=$gpg_keyring_path] $repo_url $G_DISTRO_NAME main' >> $repo_path"
```
Comment (MichaIng) [inline dietpi/dietpi-software]: Let's align the code with other 3rd party repo installs:
```suggestion
			# Detect distro
			local distro='debian'
			(( $G_HW_MODEL < 10 )) && (( $G_RASPBIAN )) && distro='raspbian'

			# APT key
			local url='https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/gpg.5AEE46706CF0453E.key'
			G_CHECK_URL "$url"
			G_EXEC eval "curl -sSfL '$url' | gpg --dearmor -o /etc/apt/trusted.gpg.d/dietpi-moonlight.gpg --yes"

			# APT list
			G_EXEC eval "echo 'deb https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/$distro/ $G_DISTRO_NAME stable' > /etc/apt/sources.list.d/dietpi-moonlight.list"
```
- I'm pretty sure there is a "debian" repo path if Debian is used instead of Raspbian, which enables ARMv8 and x86_64 support.
- Using `signed-by` instead of `trusted.gpg.d` is actually the modern/better way, but I'd like to keep things aligned for now and migrate all 3rd party repos together at some point.
- Skip source package list `deb-src`, not needed for installing binary packages. If really someone wants to compile it manually, its trivial to add it manually.

File: dietpi/dietpi-software [hunk 4]
```diff
@@ -12689,6 +12723,15 @@ If no WireGuard (auto)start is included, but you require it, please do the follo
 			[[ -d '/usr/lib/x86_64-linux-gnu' ]] && G_EXEC rmdir --ignore-fail-on-non-empty /usr/lib/x86_64-linux-gnu
 		fi
 
+		if To_Uninstall 206 # Moonlight
+		then
+			G_AGP moonlight-embedded
+			local gpg_keyring_path="/usr/share/keyrings/moonlight-game-streaming-moonlight-embedded-archive-keyring.gpg"
+			local repo_path="/etc/apt/sources.list.d/moonlight-game-streaming-moonlight-embedded.list"
+			[[ -f "$gpg_keyring_path" ]] && G_EXEC rm "$gpg_keyring_path"
+			[[ -f "$repo_path" ]] && G_EXEC rm "$repo_path"
```
Comment (MichaIng) [inline dietpi/dietpi-software]: I'm no fans of defining variables which are used only 1-2 time 😉:
```suggestion
			[[ -f '/etc/apt/sources.list.d/dietpi-moonlight.list' ]] && G_EXEC rm /etc/apt/sources.list.d/dietpi-moonlight.list
			[[ -f '/etc/apt/trusted.gpg.d/dietpi-moonlight.gpg' ]] && G_EXEC rm /etc/apt/trusted.gpg.d/dietpi-moonlight.gpg
```

**Comment by MichaIng** [comment]
Check out how we enable fake KMS for Amiberry:
```sh
/boot/dietpi/func/dietpi-set_hardware rpi-opengl vc4-fkms-v3d
```
IMO no need to revert this on uninstall since we do not know whether legacy graphics or full KMS is wanted. Fake KMS is actually quite fine for compatibility: DRM/KMS is available but some legacy graphics APIs as well (like DispmanX).

File: dietpi/dietpi-software (hunk 2 above)
Comment (mtekman) [reply dietpi/dietpi-software]: Ah nice, I didn't know that about the keyrings. I thought it was how developers signed their packages.

> apt-transport-https, ca-certificates, gnupg

Yeah I figured, wasn't sure whether to be explicit or not

File: dietpi/dietpi-software (hunk 3 above)
Comment (mtekman) [reply dietpi/dietpi-software]: Makes sense!

File: dietpi/dietpi-software (hunk 4 above)
Comment (mtekman) [reply dietpi/dietpi-software]: agreed!

File: dietpi/dietpi-software [hunk 5]
```diff
@@ -11103,6 +11109,27 @@ _EOF_
 			modprobe binfmt_misc 2> /dev/null && G_EXEC systemctl restart systemd-binfmt
 		fi
 
+		if To_Install 206 # Moonlight
+		then
+			## Install script adapted from
+			## - https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/setup.deb.sh
+			# Detect distro
+			local distro='debian'
+			(( $G_HW_MODEL < 10 )) && (( $G_RASPBIAN )) && distro='raspbian'
```
Comment (MichaIng) [inline dietpi/dietpi-software]: As of: https://github.com/MichaIng/DietPi/pull/6303#issuecomment-1509859764
```suggestion
			# Enable required fake KMS driver on RPi
			(( $G_HW_MODEL > 9 )) || /boot/dietpi/func/dietpi-set_hardware rpi-opengl vc4-fkms-v3d
```

File: dietpi/dietpi-software [hunk 6]
```diff
@@ -11113,20 +11113,17 @@ _EOF_
...
-			G_EXEC eval "curl -1sLf '$gpg_url' | gpg --dearmor > '$gpg_keyring_path'"
+			local url='https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/gpg.5AEE46706CF0453E.key'
+			G_CHECK_URL "$url"
+			G_EXEC eval "curl -sSfL '$url' | gpg --dearmor -o /etc/apt/trusted.gpg.d/dietpi-moonlight.gpg --yes"
 
 			# APT list
-			local repo_url="$base_url/deb/raspbian"
-			G_EXEC eval "echo 'deb [signed-by=$gpg_keyring_path] $repo_url $G_DISTRO_NAME main' > $repo_path"
-			G_EXEC eval "echo 'deb-src [signed-by=$gpg_keyring_path] $repo_url $G_DISTRO_NAME main' >> $repo_path"
+			G_EXEC eval "echo 'deb https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/$distro/ $G_DISTRO_NAME stable' > /etc/apt/sources.list.d/dietpi-moonlight.list"
```
Comment (MichaIng) [inline dietpi/dietpi-software]:
```suggestion
			G_EXEC eval "echo 'deb https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/deb/raspbian/ $G_DISTRO_NAME main' > /etc/apt/sources.list.d/dietpi-moonlight.list"
```

**Comment by MichaIng** [comment]
Okay, while there is a "debian" path for this repo an both have suites for all architectures, everything is empty aside of the "raspbian" armhf and arm64 ones. So it is ARM-only and probably even only for Raspberry Pi. However let's adjust it and test on other ARM SBCs.

File: dietpi/dietpi-software [hunk 7]
```diff
@@ -11103,6 +11109,27 @@ _EOF_
 			modprobe binfmt_misc 2> /dev/null && G_EXEC systemctl restart systemd-binfmt
 		fi
 
+		if To_Install 206 # Moonlight
+		then
+			## Install script adapted from
+			## - https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/setup.deb.sh
+			# Detect distro
+			local distro='debian'
```
Comment (MichaIng) [inline dietpi/dietpi-software]:
```suggestion
			# https://dl.cloudsmith.io/public/moonlight-game-streaming/moonlight-embedded/setup.deb.sh

```

File: dietpi/dietpi-software [hunk 8]
```diff
@@ -966,6 +966,12 @@ Available commands:
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,1]=0
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,2]=0
 		aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,10]=0
+		#------------------
+		software_id=206
+		aSOFTWARE_NAME[$software_id]='Moonlight'
+		aSOFTWARE_DESC[$software_id]='Game streaming client for Sunshine and NVIDIA GameStream'
+		aSOFTWARE_CATX[$software_id]=5
+		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/gaming/#moonlight'
```
Comment (MichaIng) [inline dietpi/dietpi-software:983]:
```suggestion
		# ARM only
		(( $G_HW_ARCH > 3 )) && aSOFTWARE_AVAIL_G_HW_ARCH[$software_id,$G_HW_ARCH]=0
		# No Bookworm packages yet and Bullseye packages depend on libssl1.1
		aSOFTWARE_AVAIL_G_DISTRO[$software_id,7]=0

```

**Comment by MichaIng** [comment]
I cannot test actual streaming, but the CLI at least works fine on e.g. NanoPi R5S (Debian Bullseye ARMv8).

**Comment by mtekman** [comment]
Nice, thanks for checking the upstream for other architectures!

I will test this PR on my device on Wednesday/Thursday night

**Comment by MichaIng** [comment]
About non-RPi SBCs in general: We should test it at least. This repo really seems to be made for RPi only and as it explicitly states to require the fake KMS video driver, it is possibly build against the related `libraspberrypi0` libraries which do not exist on other SBCs.

I found Sunshine, the streaming host software, which is open source as well and has Debian packages provided along with the releases: https://github.com/LizardByte/Sunshine
Would be great if someone finds time to test connecting this with Moonlight on any non-RPi SBC. I hope to find time for this as well soon. The problem is that I have no spare x86 system with the minimum requirements they list. I hope those are no hard requirements but more for fluid streaming performance reasons.

_Sunshine would be actually a nice addition in our software catalogue as well 🙂._

**Comment by mtekman** [comment]
If portability becomes a real issue, we could try adding [moonlight-qt](https://github.com/moonlight-stream/moonlight-docs/wiki/Installing-Moonlight-Qt-on-Raspberry-Pi-4) as a desktop app. I think it builds against more targets than moonlight-embedded.

I'm glad you mentioned Sunshine, as I use it religiously and it is far far better than the original Nvidia Gamestream in terms of performance and latency. So far it works flawlessly between my Desktop PC and my android smartphone (running Moonlight), and is even fantastic for full Desktop streaming.

It was definitely going to be the next package I was going to consider adding!

**Comment by MichaIng** [comment]
I think generally Moonlight it can compile on any SBC and even x86. It supports SDL as graphics backend as last resort. So it's just about the package repository which likely supports RPi only as it is compiled to use DispmanX. Of course we could add the compiling steps for other hardware of provide own packages, but that would be a larger effort to implement. Since most DietPi systems are RPi's anyway, I think it's fine to start with this first and in case add support for more hardware after we got some testing/feedback/polishing with these pre-compiled packages.

**Comment by MichaIng** [comment]
Jep, those packages are RPi-only:
```console
root@OrangePi5:~# moonlight stream -app Desktop 192.168.1.10
Connecting to 192.168.1.10...
libcec.so.6: cannot open shared object file: No such file or directory
libcec.so.6: cannot open shared object file: No such file or directory
Failed to initialize libcec interface
RTSP port: 48010
Initializing platform...done
Resolving host name...done
Initializing audio stream...done
Starting RTSP handshake...Audio port: 48000
Video port: 47998
Control port: 47999
done
Initializing control stream...done
Initializing video stream...done
Initializing input stream...done
Starting control stream...done
Starting video stream...* failed to open vchiq instance
```
Additionally, libcec is required, or more precisely tried to be loaded. I'll add it to dependencies.

**Comment by MichaIng** [comment]
Looks good from my end, but I have no capable RPi for testing actual streaming. Pairing with Sunshine from Windows desktop however works fine.

Opened an issue for adding docs: https://github.com/MichaIng/DietPi-Docs/issues/873

**Comment by mtekman** [comment]
Literally testing it right now -- for some reason, it finds the stream but refuses to display it because it can't initialize the CEC device.

Under the settings of:

```
dtoverlay=vc4-fkms-v3d
dtoverlay=rpivid-v4l2
```

and a `cec-client -l` of:

```
libCEC version: 6.0.2, compiled on Linux-5.4.0-72-generic ... , features: P8_USB, DRM, P8_detect, randr, RPi, Exynos, Linux, AOCEC
Found devices: 1

device:              1
com port:            RPI
vendor id:           2708
product id:          1001
firmware version:    1
```

If I try a simple stream of `moonlight stream -app "Desktop" -windowed -verbose`, I get:

```
Searching for server...
Connecting to 192.168.178.20...
GPU: , GFE: 3.23.0.74 (, 7.1.431.-1)
Platform Raspberry Pi (Broadcom) MMAL
Loading mappingfile /usr/share/moonlight/gamecontrollerdb.txt
Not mapping 2.4G Mouse as a gamepad
Not mapping 2.4G Mouse as a gamepad
Failed to initialize libcec interface
Stream 1280 x 720, 60 fps, 10000 kbps
RTSP port: 48010
Initializing platform...done
Resolving host name...done
Initializing audio stream...done
Starting RTSP handshake...Audio port: 48000
Video port: 47998
Control port: 47999
done
Initializing control stream...done
Initializing video stream...done
Initializing input stream...done
Starting control stream...done
Starting video stream...done
Starting audio stream...done
Starting input stream...done
Received first video packet after 300 ms
```

where nothing happens. It apparently cannot find the CEC device (whatever that is).

The odd thing is that `moonlight-qt` appears to run absolutely fine with these settings from the terminal.  Audio great, framerate great (720p @60fps), absolutely fine from the QT app.

I just can't get the standalone moonlight to work. Possible related: https://github.com/moonlight-stream/moonlight-embedded/issues/828


### Plan change:

We build `moonlight-qt` instead? 
I believe this is what I actually used before on my Pi4 and mistakenly thought that `moonlight` would be the same

**Comment by MichaIng** [comment]
CEC is a HDMI feature to control the HDMI device: https://en.wikipedia.org/wiki/Consumer_Electronics_Control
It is however not supported by all HDMI displays and shouldn't be required. Purging libcec again does not make a difference, does it?
```sh
apt autopurge libcec6
```
Did you reboot after the install so fKMS is really enabled?

Ah, what I missed is to as well raise GPU memory to defaults. Please try this before rebooting:
```sh
/boot/dietpi/func/dietpi-set_hardware gpumemsplit 76
reboot
```

**Comment by mtekman** [comment]
I just purged, set to fpkms,  and did the gpumem split to 76, then rebooted.

```
libcec.so.6: cannot open shared object file: No such file or directory
libcec.so.6: cannot open shared object file: No such file or directory
Failed to initialize libcec interface
````

CEC seems to be baked into the `moonlight-embedded` for the RPi... but then when I do `ldd /usr/bin/moonlight` libcec is not in the list of required libraries...

> I'd add moonlight-qt as additional/dedicated install option, not as replacement.

Will they be installed together then, or will users choose?

**Comment by MichaIng** [comment]
> CEC seems to be baked into the moonlight-embedded for the RPi... but then when I do ldd /usr/bin/moonlight libcec is not in the list of required libraries...

I found the same, even every other moonlight library does not show libcec via `ldd`. I guess it is loaded on demand. However, this does not break the streaming, does it? In my case logs just continued went on after the "missing libcec" error, so it does not seem to be required.

**Comment by mtekman** [comment]
Yeah, that's the thing -- it otherwise looks like it finds the stream and connects the mouse and keyboard inputs successfully, it's just somehow unable to find a device to render the video to (I've tried this both from the console and from X11)

Okay, so I guess another PR is in order :-)

I have both installed at the moment. One thing I want to test is whether I can have `moonlight-qt` and `kodi` running on the same overlay, but maybe that'd be more relavant for another PR

**Comment by MichaIng** [comment]
It is not ARMv6 compatible: https://github.com/moonlight-stream/moonlight-embedded/issues/867

**Comment by MichaIng** [comment]
Can you clone the branch and create a new PR? I do not want to give up on the non-X approach which should be generally preferred. I'll be able to test on RPi 2 this week, and if it fails the same way, we can report to Moonlight devs to see where the problem is or whether it is a bug.

**Comment by MichaIng** [comment]
It works here. I'm currently remote-controlling my laptop with my RPi 2, connected via Moonlight embedded 😃. Using
```sh
moonlight pair 192.168.1.14
# Typing the returned PIN into Sunshine web interface, then
moonlight stream -app=Desktop 192.168.1.14
```
Quite a nice alternative to VNC and similar, ähm, a no it's the wrong way round 😄. Still nice.

Alternatively one can modify `/etc/moonlight.conf` and run:
```sh
moonlight /etc/moonlight.conf
```
Not sure why it does not load this pre-installed config file automatically  🤔. However, with this, we should add a nice systemd service and an autostart option, so it runs with a dedicated unprivileged user, loads the default config files and optionally can be started automatically at boot and/or after login.

**Review by MichaIng — APPROVED** [review]

**Comment by MichaIng** [comment]
More testers are however highly welcome 🙂.

**Comment by MichaIng** [comment]
@mtekman 
I just found: https://nvidia.custhelp.com/app/answers/detail/a_id/5436/~/gamestream-end-of-service-notification
Does this mean the info that Moonlight is a client for NVIDIA GameStream is obsolete, as it doesn't exist anymore, and instead Sunshine is the only streaming server option now? Or are there other compatible streaming server solutions?

Reminds me that [Steam Link](https://github.com/MichaIng/DietPi/pull/6303) was another software requests which got quite many votes on FeatHub. Since Moonlight also supports Steam when connected to a Sunshine instance (but I didn't test this yet), we have at least an alternative with this combination now.

**Comment by mtekman** [comment]
AFAIK, Moonlight is still compatible with the Gamestream protocol as it was before the end of service announcement. So people who didn't upgrade their GeForce Experience should still be able to stream to Moonlight using GameStream. I don't think the protocol was too hard to re-implement and so Sunshine was born to get around this.

Nvidia are now pushing hard their shield tablets which get streams from their own cloud computing service, so I think that whatever new streaming protocol they're implementing (probably with some encryption baked in) probably won't be such an easily understandable protocol like GameStream was, meaning Moonlight probably won't be compatible with anything Nvidia are doing on the cloud right now (?)

As for SteamLink, it looks like they're using something called [RemotePlay](https://store.steampowered.com/remoteplay/) on the server side of things, and are picking it up on the client side via SteamLink apps. Moonlight is apparently [implemented as an app in Steamlink](https://steamcommunity.com/app/353380/discussions/1/1814296273137162419/), but it's not clear if Moonlight itself can use RemotePlay streams.

I found this [one single mention](https://github.com/moonlight-stream/moonlight-android/issues/411#issuecomment-352070303) of using Moonlight to pick up RemotePlay streams, but it's still not clear to me if this is something that Moonlight actually supports.

TL;DR - Yes, Sunshine is realistically the main open streaming server solution for gaming right now, and Moonlight is the main client, which potential support for RemotePlay

This is just my personal perspective on this, I could be dead wrong.

**Comment by MichaIng** [comment]
> So people who didn't upgrade their GeForce Experience should still be able to stream to Moonlight using GameStream.

As far as I understand the [Moonlight](https://github.com/moonlight-stream/moonlight-docs/wiki/NVIDIA-GameStream-End-Of-Service-Announcement-FAQ) wiki entry about this, it is not possible anymore to use even older NVIDIA GameStream apps.

However, of course Moonlight works with Sunshine independently and other GameStream compatible streaming host servers are possioble, I was more asking since we explicitly name "NVIDIA GameStream" in the install options description, which is weird if this service is dead.
